File: c_implementation/linked_list.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// Node structure representing each element in the linked list
typedef struct Node {
    int data;              // Data stored in the node
    struct Node* next;     // Pointer to the next node
} Node;

// LinkedList structure representing the linked list
typedef struct LinkedList {
    Node* head;            // Pointer to the head node
} LinkedList;

// Creates and initializes a new linked list
LinkedList* create_linked_list() {
    LinkedList* list = (LinkedList*)malloc(sizeof(LinkedList));
    list->head = NULL; // Initialize head to NULL
    return list;
}

// Checks if the linked list is empty
bool is_empty(LinkedList* list) {
    return list->head == NULL;
}

// Returns the size (number of nodes) of the linked list
int size(LinkedList* list) {
    int count = 0;
    Node* current = list->head;
    while (current != NULL) {
        count++;
        current = current->next;
    }
    return count;
}

// Adds a new node with the given data at the head of the list
void add(LinkedList* list, int data) {
    Node* new_node = (Node*)malloc(sizeof(Node));
    new_node->data = data;
    new_node->next = list->head;
    list->head = new_node;
}

// Searches for a node with the given key and returns it
Node* search(LinkedList* list, int key) {
    Node* current = list->head;
    while (current != NULL) {
        if (current->data == key) {
            return current; // Return the node if found
        }
        current = current->next;
    }
    return NULL; // Return NULL if not found
}

// Inserts a new node with the given data at the specified index
void insert(LinkedList* list, int data, int index) {
    if (index == 0) {
        add(list, data); // Add at the head if index is 0
        return;
    }

    Node* new_node = (Node*)malloc(sizeof(Node));
    new_node->data = data;

    Node* current = list->head;
    for (int i = 0; i < index - 1 && current != NULL; i++) {
        current = current->next;
    }

    if (current == NULL) {
        printf("Index out of bounds\n");
        free(new_node);
        return;
    }

    new_node->next = current->next;
    current->next = new_node;
}

// Removes the node with the given key and returns its data
int remove_node(LinkedList* list, int key) {
    Node* current = list->head;
    Node* previous = NULL;

    while (current != NULL) {
        if (current->data == key) {
            if (previous == NULL) {
                list->head = current->next; // Update head if removing the first node
            } else {
                previous->next = current->next;
            }
            int data = current->data;
            free(current);
            return data; // Return the data of the removed node
        }
        previous = current;
        current = current->next;
    }
    return -1; // Return -1 if key not found
}

// Prints the linked list in a readable format
void print_list(LinkedList* list) {
    Node* current = list->head;
    while (current != NULL) {
        printf("[%d] -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");
}

int main() {
    LinkedList* list = create_linked_list();

    // Add nodes to the list
    add(list, 10);
    add(list, 20);
    add(list, 30);

    printf("List: ");
    print_list(list);

    printf("Size: %d\n", size(list));

    // Search for a node
    Node* found = search(list, 20);
    if (found) {
        printf("Found: %d\n", found->data);
    } else {
        printf("Not found\n");
    }

    // Insert a node at index 1
    insert(list, 25, 1);
    printf("After insertion: ");
    print_list(list);

    // Remove a node with key 20
    remove_node(list, 20);
    printf("After removal: ");
    print_list(list);

    return 0;
}

File: linked_list.py
class Node:
    """
    An object for storing a single node of a linked list.
    Models two attributes - data and the link to the next node in the list.
    """
    def __init__(self, data):
        self.data = data          # Assign data when a new node is created
        self.next_node = None     # This points to the next node in the list

    def __repr__(self):
        return "<Node data: %s>" % self.data  # Display node in a readable format


class LinkedList:
    # Singly linked list

    def __init__(self):
        self.head = None
        self.__count = 0           # Keep track of number of nodes for O(1) size

    def is_empty(self):
        # This checks if head = None (i.e. if the list is empty)
        return self.head is None

    def size(self):
        """
        Returns the number of nodes in the list.
        Takes O(1) time due to tracking with __count.
        """
        return self.__count

    def add(self, data):
        """
        Adds new node containing data at head.
        Takes O(1) time.
        """
        new_node = Node(data)
        new_node.next_node = self.head
        self.head = new_node
        self.__count += 1

    def search(self, key):
        """
        Search for the first node containing data that matches the key.
        Returns the node or `None` if not found.
        Takes O(n) time.
        """
        current = self.head

        while current:
            if current.data == key:
                return current
            current = current.next_node
        return None

    def insert(self, data, index):
        """
        Inserts a new Node containing data at index position.
        Insertion takes O(1) time but finding node at insertion point takes
        O(n) time. Overall O(n).
        """
        if index > self.__count or index < 0:
            raise IndexError('Index out of range')

        if index == 0:
            self.add(data)
            return

        new_node = Node(data)
        current = self.head
        position = 0

        # Traverse to node just before the insertion point
        while position < index - 1:
            current = current.next_node
            position += 1

        new_node.next_node = current.next_node
        current.next_node = new_node

        self.__count += 1

    def node_at_index(self, index):
        """
        Returns the Node at specified index.
        Takes O(n) time.
        """
        if index >= self.__count or index < 0:
            raise IndexError('Index out of range')

        current = self.head
        position = 0

        while position < index:
            current = current.next_node
            position += 1

        return current

    def remove(self, key):
        """
        Removes Node containing data that matches the key.
        Returns the node or `None` if key doesn't exist.
        Takes O(n) time.
        """
        current = self.head
        previous = None

        while current:
            if current.data == key:
                if previous is None:  # Node to remove is head
                    self.head = current.next_node
                else:
                    previous.next_node = current.next_node
                self.__count -= 1
                return current
            previous = current
            current = current.next_node

        return None

    def remove_at_index(self, index):
        """
        Removes Node at specified index.
        Takes O(n) time.
        """
        if index >= self.__count or index < 0:
            raise IndexError('Index out of range')

        current = self.head

        if index == 0:
            self.head = current.next_node
            self.__count -= 1
            return current

        position = 0
        previous = None

        while position < index:
            previous = current
            current = current.next_node
            position += 1

        previous.next_node = current.next_node
        self.__count -= 1

        return current

    def __iter__(self):
        current = self.head
        while current:
            yield current
            current = current.next_node

    def __repr__(self):
        """
        Returns a string representation of the list.
        """
        nodes = []
        current = self.head

        while current:
            if current is self.head:
                nodes.append("[Head: %s]" % current.data)
            elif current.next_node is None:
                nodes.append("[Tail: %s]" % current.data)
            else:
                nodes.append("[%s]" % current.data)

            current = current.next_node

        return ' -> '.join(nodes)


# Example usage:
if __name__ == "__main__":
    l = LinkedList()  # create a new list

    print("Is the list empty?", l.is_empty())  # expected: True
    print("Size of the list:", l.size())       # expected: 0

    l.add(5)
    print("Is the list empty?", l.is_empty())  # expected: False
    print("Size of the list:", l.size())       # expected: 1
    print("List:", l)

    l.add(10)
    l.add(15)
    print("List after adding 10 and 15:", l)

    n = l.search(10)
    print("Search for 10:", n)

    l.insert(7, 1)
    print("List after inserting 7 at index 1:", l)

    removed_node = l.remove(15)
    print("Removed node with data 15:", removed_node)
    print("List after removal:", l)

    removed_at_index = l.remove_at_index(1)
    print("Removed node at index 1:", removed_at_index)
    print("List after removing at index 1:", l)

    print("Iterating through list:")
    for node in l:
        print(node)
